core/trino-main/src/main/java/io/trino/sql/SessionSpecificationEvaluator.java

```diff
@@ -0,0 +1,208 @@
+                }
+                sessionProperties.put(nameParts.getFirst(), toSessionValue(session, parameters, specification, systemSessionPropertyMetadata.get()));
+            }
+
+            if (nameParts.size() == 2) {
+                CatalogHandle catalogHandle = getRequiredCatalogHandle(plannerContext.getMetadata(), session, specification, nameParts.getFirst());
+                Optional<PropertyMetadata<?>> connectorSessionPropertyMetadata = sessionPropertyManager.getConnectorSessionPropertyMetadata(catalogHandle, nameParts.getLast());
+                if (connectorSessionPropertyMetadata.isEmpty()) {
+                    throw semanticException(INVALID_SESSION_PROPERTY, specification, "Session property %s does not exist", specification.getName());
+                }
+                catalogProperties.put(nameParts.get(0), nameParts.get(1), toSessionValue(session, parameters, specification, connectorSessionPropertyMetadata.get()));
+            }
+        }
+
+        return new ResolvedSessionSpecifications(sessionProperties.buildOrThrow(), catalogProperties.rowMap());
+    }
+
+    public Session overrideProperties(Session session, ResolvedSessionSpecifications resolvedSessionSpecifications)
+    {
+        requireNonNull(resolvedSessionSpecifications, "resolvedSessionSpecifications is null");
+
+        validateSystemProperties(session, resolvedSessionSpecifications.systemProperties());
+
+        // Catalog session properties were already evaluated so we need to evaluate overrides
+        if (session.getTransactionId().isPresent()) {
+            validateCatalogProperties(session, resolvedSessionSpecifications.catalogProperties());
+        }
+
+        // NOTE: properties are validated before calling overrideProperties
+        Map<String, String> systemProperties = new HashMap<>();
+        systemProperties.putAll(session.getSystemProperties());
+        systemProperties.putAll(resolvedSessionSpecifications.systemProperties());
+
+        Map<String, Map<String, String>> catalogProperties = new HashMap<>(session.getCatalogProperties());
+        for (Map.Entry<String, Map<String, String>> catalogEntry : resolvedSessionSpecifications.catalogProperties().entrySet()) {
+            catalogProperties.computeIfAbsent(catalogEntry.getKey(), id -> new HashMap<>())
+                    .putAll(catalogEntry.getValue());
+        }
+
+        return session.withProperties(systemProperties, catalogProperties);
+    }
+
+    private String toSessionValue(Session session, Map<NodeRef<Parameter>, Expression> parameters, SessionSpecification specification, PropertyMetadata<?> propertyMetadata)
+    {
+        Type type = propertyMetadata.getSqlType();
+        Object objectValue;
+
+        try {
+            objectValue = evaluatePropertyValue(specification.getValue(), type, session, plannerContext, accessControl, parameters);
```

Comment:
Consider extracting common code from here and `SetSessionTask`

Reply:
The logic is a little different from `SetSessionTask`. Added a TODO comment.